### ISSUE TITLE
Wrong dst() value for time in Portugal tz

### ISSUE THREAD

**Issue by abalkin** [dateutil/dateutil#128]
Pick a time in the Summer of 1996:
```
>>> t = datetime(1996, 6, 1, tzinfo=tz.gettz('Portugal'))
```
the time zone abbreviation is computed correctly:
```
>>> t.tzname()
'WEST'
```
([WEST](http://www.timeanddate.com/time/zones/west) = Western European Summer Time).

However, 

```
>>> t.dst()
datetime.timedelta(0)
```

and as a consequence:

```
>>> t.timetuple().tm_isdst
0
>>> from time import strftime
>>> strftime('%Z', t.timetuple())
'WET'
```

The problem is that `t.dst()` is computed as a difference between `t.utcoffset()` and the offset of the last standard time before `t`, but Portugal switched to `WEST` in the Spring of 1996 from `CET`, not `WET` and therefore UTC offset did not change, while isdst did.

**Comment by pganssle** [comment]
@abalkin Is there actually enough information in the `tzfile` to solve this problem other than by using the dubious property of changes in the abbreviation? I know the `tzdata` inputs have a "SAVE" column, but that seems to be removed in the `zic`.

I suppose that the "next best solution" is to make two assumptions:

1. There is no real zone where the DST offset is == 0
2. There are no zones where both the DST offset and the base zone changed in  *both* the STD->DST transition *and* the DST->STD transition.

In which case, if we know it's DST from the `isdst` flag, and the calculated `dst()` is 0, then we should calculate the DST offset from the difference in offset with the _next_ standard time.

This still leaves us open to potential zones where the base zone change goes the other way (making the `dst()` value 2 hours instead of 1).

**Comment by abalkin** [comment]
>  Is there actually enough information in the tzfile to solve this problem other than by using the dubious property of changes in the abbreviation?

Yes, moreover, lossless UTC-local conversion can be implemented using POSIX localtime_r API and without mktime or equivalent.  See python/cpython#20.  (Note that only pure python implementation is ready at this point.)

**Comment by abalkin** [comment]
@pganssle - sorry, I thought I was responding on #112.  In this case, you are right: the necessary information is only available in raw files and is not preserved by zic.

**Comment by abalkin** [comment]
@pganssle - please see my [Computing .dst() as a timedelta](https://mail.python.org/pipermail/datetime-sig/2015-September/000882.html) post on the Datetime-SIG mailing list.

**Comment by pganssle** [comment]
@abalkin I was cleaning up in the `tzfile` implementation anyway when trying to get a handle on #112, so I implemented the above solution. The behavior is now consistent with `pytz`'s behavior, at least (as far as I can tell), though we're subject to the same bug in the Portugal zone in 1940.

I don't think there's much to be done about it, though. A heuristic that covers that use case would be to check if the dst() values calculated based on the STD->DST transition is different from the DST->STD transition, and if it is, check it against the next zone transition boundary, and if that one matches one of the two transitions, fill in the uncertain one with the more popular value.

In the end, it would be nice to get the SAVE values directly from tzdata, but I think generally people will prefer to use the compiled binaries stored in the OS, so it's quite a lot of work for such a rare edge case.

**Comment by pganssle** [comment]
I will note that this will be a much bigger problem if one of these transitions happens a second time

**Comment by pganssle** [comment]
Fixed in #225.
